azurerm/resource_arm_firewall_application_rule_collection.go

```diff
@@ -0,0 +1,448 @@
+		}
+
+		if priority := props.Priority; priority != nil {
+			d.Set("priority", int(*priority))
+		}
+
+		flattenedRules := flattenFirewallApplicationRuleCollectionRules(props.Rules)
+		if err := d.Set("rule", flattenedRules); err != nil {
+			return fmt.Errorf("Error setting `rule`: %+v", err)
+		}
+	}
+
+	return nil
+}
+
+func resourceArmFirewallApplicationRuleCollectionDelete(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).azureFirewallsClient
+	ctx := meta.(*ArmClient).StopContext
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resourceGroup := id.ResourceGroup
+	firewallName := id.Path["azureFirewalls"]
+	name := id.Path["applicationRuleCollections"]
+
+	azureRMLockByName(firewallName, azureFirewallResourceName)
+	defer azureRMUnlockByName(firewallName, azureFirewallResourceName)
+
+	firewall, err := client.Get(ctx, resourceGroup, firewallName)
+	if err != nil {
+		if utils.ResponseWasNotFound(firewall.Response) {
+			// assume deleted
+			return nil
+		}
+
+		return fmt.Errorf("Error making Read request on Azure Firewall %q (Resource Group %q): %+v", firewallName, resourceGroup, err)
+	}
+
+	props := firewall.AzureFirewallPropertiesFormat
+	if props == nil {
+		return fmt.Errorf("Error retrieving Application Rule Collection %q (Firewall %q / Resource Group %q): `props` was nil", name, firewallName, resourceGroup)
+	}
+	if props.ApplicationRuleCollections == nil {
+		return fmt.Errorf("Error retrieving Application Rule Collection %q (Firewall %q / Resource Group %q): `props.ApplicationRuleCollections` was nil", name, firewallName, resourceGroup)
+	}
+
+	applicationRules := make([]network.AzureFirewallApplicationRuleCollection, 0)
+	for _, rule := range *props.ApplicationRuleCollections {
+		if rule.Name == nil {
+			continue
+		}
+
+		if *rule.Name != name {
+			applicationRules = append(applicationRules, rule)
+		}
+	}
+	props.ApplicationRuleCollections = &applicationRules
+
+	future, err := client.CreateOrUpdate(ctx, resourceGroup, firewallName, firewall)
+	if err != nil {
+		return fmt.Errorf("Error deleting Application Rule Collection %q from Firewall %q (Resource Group %q): %+v", name, firewallName, resourceGroup, err)
+	}
+
+	err = future.WaitForCompletionRef(ctx, client.Client)
+	if err != nil {
+		return fmt.Errorf("Error waiting for deletion of Application Rule Collection %q from Firewall %q (Resource Group %q): %+v", name, firewallName, resourceGroup, err)
+	}
+
+	return nil
+}
+
+func expandArmFirewallApplicationRules(inputs []interface{}) ([]network.AzureFirewallApplicationRule, error) {
+	outputs := make([]network.AzureFirewallApplicationRule, 0)
+
+	for _, input := range inputs {
+		rule := input.(map[string]interface{})
+
+		ruleName := rule["name"].(string)
+		ruleDescription := rule["description"].(string)
+
+		ruleSourceAddresses := make([]string, 0)
+		for _, v := range rule["source_addresses"].(*schema.Set).List() {
```

Comment:
We have a function that handles this common functionality: `ExpandStringArray`. All these loops could be replaced with it i believe.

Reply:
You could also write a complimentary `set.ToStringSlice(i interface{}{})` function that explicitly knows about sets.